File: LeetCode/2940. Find Building Where Alice and Bob Can Meet/solution.py
# https://leetcode.com/problems/find-building-where-alice-and-bob-can-meet/solutions/4304269/java-c-python-priority-queue/
class Solution:
    def leftmostBuildingQueries(self, heights: List[int], queries: List[List[int]]) -> List[int]:
        def ignore(i,j):
            return i == j or heights[i] < heights[j]
        n, m = len(heights), len(queries)
        ans = [-1]*m 
        candidates = [[] for _ in range(n)]
        for i, (l,r) in enumerate(queries):
            if l > r: l, r = r, l  
            if ignore(l,r):
                ans[i] = r 
            else:
                candidates[r].append((max(heights[l],heights[r]),i))
        hq = []
        for r, h in enumerate(heights):
            while hq and hq[0][0] < h:
                ans[hq[0][-1]] = r 
                heappop(hq)
            for hx, i in candidates[r]:
                heappush(hq,(hx,i))
        return ans 

class Solution:
    def leftmostBuildingQueries(self, heights: List[int], queries: List[List[int]]) -> List[int]:
        n = len(heights)
        m = len(queries)
        idx = list(range(m))
        idx.sort(key = lambda x: -max(queries[x]))
        ans = [-1]*m 
        j = n-1
        st = []
        for i in idx:
            l, r = queries[i]
            if l > r : l, r = r, l 
            if l == r: 
                ans[i] = l
                continue
            while j >= r:
                while st and heights[st[-1]] < heights[j]:
                    st.pop()
                st.append(j)
                j -= 1
            lo, hi = 0, len(st)-1
            while lo <= hi:
                t = (lo+hi)//2
                if heights[st[t]] > heights[l]:
                    lo = t+1
                else:
                    hi = t-1
            if hi >= 0:
                ans[i] = st[hi]
        return ans 

# 在线查询+st表+二分
class Solution:
    def leftmostBuildingQueries(self, heights: List[int], queries: List[List[int]]) -> List[int]:
        n, m = len(heights), len(queries)
        ans = [-1]*m 
        sz = 0
        while (1<<sz) <= n:
            sz += 1
        st = [[0]*sz for _ in range(n)]
        for i in range(n): st[i][0] = heights[i]
        for j in range(1,sz):
            for i in range(n):
                if i+(1<<j) > n: break 
                st[i][j] = max(st[i][j-1],st[i+(1<<(j-1))][j-1])
        log2 = [0]*n 
        j = 0
        for i in range(1,n):
            if (1<<(j+1)) < i+1: j += 1
            log2[i] = j 
        # print(st)
        # print(log2)
        def query(l,r):
            j = log2[r-l+1]
            d = 1<<j
            return max(st[l][j],st[r-d+1][j])
        for i, (l, r) in enumerate(queries):
            # print("query",i,l,r)
            if l > r: l, r = r, l 
            if l == r or heights[l] < heights[r]:
                ans[i] = r 
                continue
            lo, hi = r, n-1
            while lo <= hi:
                t = (lo+hi)//2
                # print(r,t,query(r,t))
                if query(r,t) > heights[l]:
                    hi = t-1
                else:
                    lo = t+1
            if lo < n:
                ans[i] = lo 
        return ans 

File: LeetCode/2940. Find Building Where Alice and Bob Can Meet/solution.cpp
class Solution {
public:
    vector<int> leftmostBuildingQueries(vector<int>& heights, vector<vector<int>>& queries) {
        int n = heights.size();

        vector<array<int, 3>> vec;
        int q = queries.size();
        vector<int> ans(q, -2);
        for (int i = 0; i < q; i++) {
            int a = min(queries[i][0], queries[i][1]), b = max(queries[i][0], queries[i][1]);
            // 处理两种特殊询问
            if (a == b) ans[i] = a;
            else if (heights[a] < heights[b]) ans[i] = b;
            // 剩下的询问加入二维偏序问题
            // 注意排序方式：首先按 heights 从大到小排序，相同 heights 的询问一定要排在数据点之后
            else vec.push_back({-heights[a] - 1, i, b});
        }
        for (int i = 0; i < n; i++) vec.push_back({-heights[i], -1, i});
        sort(vec.begin(), vec.end());

        // 树状数组模板

        const int INF = 1e9;
        int tree[n + 1];
        for (int i = 1;i <= n; i++) tree[i] = INF;

        auto lowbit = [&](int x) { return x & (-x); };

        auto modify = [&](int pos, int val) {
            for (; pos <= n; pos += lowbit(pos)) tree[pos] = min(tree[pos], val);
        };

        auto query = [&](int pos) {
            int ret = INF;
            for (; pos; pos -= lowbit(pos)) ret = min(ret, tree[pos]);
            return ret;
        };

        // 利用树状数组处理二维偏序问题
        for (auto &arr : vec) {
            int tpe = arr[1], idx = arr[2];
            if (tpe == -1) modify(n - idx, idx);
            else {
                ans[tpe] = query(n - idx);
                if (ans[tpe] == INF) ans[tpe] = -1;
            }
        }
        return ans;
    }
};

作者：TsReaper
链接：https://leetcode.cn/problems/find-building-where-alice-and-bob-can-meet/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。